src/display/accessibility.js

```diff
@@ -0,0 +1,262 @@
+    }
+
+    this.#textLayer = textLayer;
+    let textChildren = textLayer.querySelectorAll(`span[role="presentation"]`);
+    if (textChildren.length === 0) {
+      return shadow(this, "_textLayerElements", null);
+    }
+
+    this.#textChildren = textChildren = Array.from(textChildren);
+    textChildren.sort(AccessibilityManager.#compareElementPositions);
+
+    return textChildren;
+  }
+
+  /**
+   * Compare the positions of two elements, it must correspond to
+   * the visual ordering.
+   *
+   * @param {HTMLElement} e1
+   * @param {HTMLElement} e2
+   * @returns {number}
+   */
+  static #compareElementPositions(e1, e2) {
+    const rect1 = e1.getBoundingClientRect();
+    const rect2 = e2.getBoundingClientRect();
+
+    const top1 = rect1.y;
+    const bot1 = rect1.y + rect1.height;
+    const mid1 = rect1.y + rect1.height / 2;
+
+    const top2 = rect2.y;
+    const bot2 = rect2.y + rect2.height;
+    const mid2 = rect2.y + rect2.height / 2;
+
+    if (mid1 <= top2 && mid2 >= bot1) {
+      return -1;
+    }
+
+    if (mid2 <= top1 && mid1 >= bot2) {
+      return +1;
+    }
+
+    const centerX1 = rect1.x + rect1.width / 2;
+    const centerX2 = rect2.x + rect2.width / 2;
+
+    return centerX1 - centerX2;
+  }
+
+  /**
+   * Function called when the text layer has finished rendering.
+   */
+  onTextLayerRendered() {
+    setTimeout(() => {
+      if (this.#textNodes.size > 0) {
+        const textChildren = this._textLayerElements;
+        for (const [id, nodeIndex] of this.#textNodes) {
+          this.#addIdToAriaOwns(id, textChildren[nodeIndex]);
+        }
+      }
+
+      for (const [element, isRemovable] of this.#waitingElements) {
+        this.addPointerInTextLayer(element, isRemovable);
+      }
+      this.#waitingElements.clear();
+    }, 0);
+  }
+
+  /**
+   * Remove an aria-owns id from a node in the text layer.
+   * @param {HTMLElement} element
+   */
+  removePointerInTextLayer(element) {
+    if (!this.#hasTextLayer) {
+      this.#waitingElements.delete(element);
+      return;
+    }
+
+    const { id } = element;
+    const nodeIndex = this.#textNodes.get(id);
+    if (nodeIndex === undefined) {
+      return;
+    }
+
+    const node = this._textLayerElements[nodeIndex];
+
+    this.#textNodes.delete(id);
+    let owns = node.getAttribute("aria-owns");
+    if (owns?.includes(id)) {
+      owns = owns
+        .split(" ")
+        .filter(x => x !== id)
+        .join(" ");
+      if (owns) {
+        node.setAttribute("aria-owns", owns);
+      } else {
+        node.removeAttribute("aria-owns");
+        node.setAttribute("role", "presentation");
+      }
+    }
+  }
+
+  #addIdToAriaOwns(id, node) {
+    if (
+      typeof PDFJSDev === "undefined" ||
+      PDFJSDev.test("!PRODUCTION || TESTING")
+    ) {
+      if (DEBUG) {
+        node.style.backgroundColor = "red";
+        setTimeout(() => {
+          node.style.backgroundColor = "";
+        }, 2000);
+      }
+    }
```

Comment:
Having to manually edit the code, in order to enable debugging, seems less than ideal.

The elements in question are always placed in the textLayer, right?
Because if so, it'd be *much better* to piggyback on the existing `textLayer=visible` debugging hash-parameter with appropriate rules added in the `web/debugger.css` file.

Reply:
Excellent idea !